# Pandas 

(If you're using the code files, please open pandas_lessons.py)

## The importance of data preprocessing

Data preprocessing (also called data wrangling, cleaning, scrubbing, etc) is the most important thing you will do with your data because it sets the stage for the analysis part of your data analysis workflow. The preprocessing you do largely depends on what kind of data you have, what sort of analysis you'll be doing with your data, and what you intend to do with the results.

Preprocessing is also a process for getting to know your data, and can answer questions such as these (and more): 

- What kind of data are you working with? 
- Is it categorical, continuous, or a mix of both? 
- What's the distribution of features in your dataset? 
- What sort of wrangling do you have to do?
- Do you have any missing data? 
- Do you need to remove missing data?
- Do you need only a subset of your data?
- Do you need more data?
- Or less?

The questions you'll have to answer are, again, dependent upon the data that you're working with, and preprocessing can be a way to figure that out.

## What is Pandas?

Pandas is by far my favorite preprocessing tool. It's a data wrangling/modeling/analysis tool that is similar to R and Excel; in fact, the DataFrame data structure in Pandas was named after the DataFrame in R. Pandas comes with several easy-to-use data structures, two of which (the `Series` and the `DataFrame`) I'll be covering here.

I'll also be covering a bunch of different wrangling tools, as well as a couple of analysis tools.

## Why Pandas?

So, why would you want to use Python, as opposed to tools like R and Excel? I like to use it because I like to keep everything in Python, from start to finish. It just makes it easier if I don't have to switch back and forth between other tools. Also, if I have to build in preprocessing as part of a production system, which I've had to do at my job, it makes sense to just do it in Python from the beginning. 

Pandas is great for preprocessing, as we'll see, and it can be easily combined with other modules from the scientific Python stack.

## Pandas data structures

Pandas has several different data structures, but we're going to talk about the `Series` and the `DataFrame`.

### The Series

The `Series` is a one-dimensional array that can hold a variety of data types, including a mix of those types. The row labels in a `Series` are collectively called the index. You can create a `Series` in a few different ways. Here's how you'd create a `Series` from a list.

In [2]:
import pandas as pd

some_numbers = [2, 5, 7, 3, 8]

series_1 = pd.Series(some_numbers)
series_1

0    2
1    5
2    7
3    3
4    8
dtype: int64

To specify an index, you can also pass in a list.

In [3]:
ind = ['a', 'b', 'c', 'd', 'e']

series_2 = pd.Series(some_numbers, index=ind)
series_2

a    2
b    5
c    7
d    3
e    8
dtype: int64

We can pull that index back out again, too, with the `.index` attribute.

In [4]:
series_2.index

Index([u'a', u'b', u'c', u'd', u'e'], dtype='object')

You can also create a `Series` with a dictionary. The keys of the dictionary will be used as the index, and the values will be used as the `Series` array.

In [5]:
more_numbers = {'a': 9, 'b': 'eight', 'c': 7.5, 'd': 6}

series_3 = pd.Series(more_numbers)
series_3

a        9
b    eight
c      7.5
d        6
dtype: object

Notice how, in that previous example, I created a `Series` with integers, a float, and a string.

### The DataFrame

The `DataFrame` is Pandas' most used data structure. It's a two and greater dimensional structure that can also hold a variety of mixed data types. It's similar to a spreadsheet in Excel or a SQL table. You can create a `DataFrame` with a few different methods. First, let's look at how to create a `DataFrame` from multiple `Series` objects.

In [6]:
combine_series = pd.DataFrame([series_2, series_3])
combine_series

a      b    c  d   e
0  2      5  7.0  3   8
1  9  eight  7.5  6 NaN

Notice how in column `b`, we have two kinds of data. If a column in a `DataFrame` contains multiple types of data, the data type (or `dtype`) of the column will be chosen to accomodate all of the data. We can look at the data types of different columns with the `.dtypes` attribute. `object` is the most general, which is what has been chosen for column `b`.

In [7]:
combine_series.dtypes

a      int64
b     object
c    float64
d      int64
e    float64
dtype: object

Another way to create a `DataFrame` is with a dictionary of lists. This is pretty straightforward:

In [8]:
data = {'col1': ['i', 'love', 'pandas', 'so', 'much'],
        'col2': ['so', 'will', 'you', 'i', 'promise']}

df = pd.DataFrame(data)
df

col1     col2
0       i       so
1    love     will
2  pandas      you
3      so        i
4    much  promise

## File I/O

It's really easy to read data into Pandas from a file. Pandas will read your file directly into a `DataFrame`. There are multiple ways to read in files, but they all work in the same way. Here's how you read in a CSV file:

In [9]:
wine = pd.read_csv('../data/wine.csv')
wine.head()

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
0  14.23        1.71  2.43        15.6        127           2.80        3.06   
1  13.20        1.78  2.14        11.2        100           2.65        2.76   
2  13.16        2.36  2.67        18.6        101           2.80        3.24   
3  14.37        1.95  2.50        16.8        113           3.85        3.49   
4  13.24        2.59  2.87        21.0        118           2.80        2.69   

   nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
0                  0.28             2.29   5.64  1.04      3.92     1065   
1                  0.26             1.28   4.38  1.05      3.40     1050   
2                  0.30             2.81   5.68  1.03      3.17     1185   
3                  0.24             2.18   7.80  0.86      3.45     1480   
4                  0.39             1.82   4.32  1.04      2.93      735   

   wine_type  
0          1  
1          1  
2          1  
3          1  
4          1

Reading in a text file is just as easy. Make sure to pass in `'\t'` to the delimiter parameter.

In [10]:
auto_mpg = pd.read_csv('../data/auto_mpg.txt', delimiter='\t')
auto_mpg.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0   18          8           307        130    3504          12.0     70   
1   15          8           350        165    3693          11.5     70   
2   18          8           318        150    3436          11.0     70   
3   16          8           304        150    3433          12.0     70   
4   17          8           302        140    3449          10.5     70   

   origin                   car_name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

## Exploring the data

Here are some different ways to explore the data we have. Let's first take a look at some of the basic characteristics of the auto_mpg dataset. You can easily find the number of rows and the number of columns a dataframe has using the `.shape` attribute.

In [11]:
auto_mpg.shape

(398, 9)

You've already seen the `head()` function, which returns the first five lines in the dataset. To grab the last 5 lines, you can use the `tail()` function:

In [12]:
auto_mpg.tail()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
393   27          4           140         86    2790          15.6     82   
394   44          4            97         52    2130          24.6     82   
395   32          4           135         84    2295          11.6     82   
396   28          4           120         79    2625          18.6     82   
397   31          4           119         82    2720          19.4     82   

     origin         car_name  
393       1  ford mustang gl  
394       2        vw pickup  
395       1    dodge rampage  
396       1      ford ranger  
397       1       chevy s-10

Getting column names from a `DataFrame` is also easy and can be done using the `.columns` attribute.

In [13]:
wine.columns

Index([u'abv', u'malic_acid', u'ash', u'alcalinity', u'magnesium', u'total_phenols', u'flavanoids', u'nonflavanoid_phenols', u'proanthocyacins', u'color', u'hue', u'dilution', u'proline', u'wine_type'], dtype='object')

Another useful thing you can do is generate some summary statistics using the `describe()` function. The `describe()` function calculates descriptive statistics like the mean, standard deviation, and quartile values for continuous and integer data that exist in your dataset. Don't worry, Pandas won't try to calculate the standard deviation of your categorical values!

In [14]:
wine.describe()

abv  malic_acid         ash  alcalinity   magnesium  \
count  178.000000  178.000000  178.000000  178.000000  178.000000   
mean    13.000618    2.336348    2.366517   19.494944   99.741573   
std      0.811827    1.117146    0.274344    3.339564   14.282484   
min     11.030000    0.740000    1.360000   10.600000   70.000000   
25%     12.362500    1.602500    2.210000   17.200000   88.000000   
50%     13.050000    1.865000    2.360000   19.500000   98.000000   
75%     13.677500    3.082500    2.557500   21.500000  107.000000   
max     14.830000    5.800000    3.230000   30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyacins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

            color         hue    dilution      proline   wine_type  
count  178.000000  178.000000  178.000000   178.000000  178.000000  
mean     5.058090    0.957449    2.611685   746.893258    1.938202  
std      2.318286    0.228572    0.709990   314.907474    0.775035  
min      1.280000    0.480000    1.270000   278.000000    1.000000  
25%      3.220000    0.782500    1.937500   500.500000    1.000000  
50%      4.690000    0.965000    2.780000   673.500000    2.000000  
75%      6.200000    1.120000    3.170000   985.000000    3.000000  
max     13.000000    1.710000    4.000000  1680.000000    3.000000

Another useful thing you can do to explore your data is to sort it. Let's say we wanted to sort our `auto_mpg DataFrame` by mpg. This is very easy as well:

In [15]:
auto_mpg.sort(columns='mpg').tail()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
326  43.4          4            90         48    2335          23.7     80   
394  44.0          4            97         52    2130          24.6     82   
325  44.3          4            90         48    2085          21.7     80   
329  44.6          4            91         67    1850          13.8     80   
322  46.6          4            86         65    2110          17.9     80   

     origin              car_name  
326       2    vw dasher (diesel)  
394       2             vw pickup  
325       2  vw rabbit c (diesel)  
329       3   honda civic 1500 gl  
322       3             mazda glc

## Lesson: let's see what's going on in our data!

This dataset is data on credit approvals. The column names and data were changed to protect the confidentiality of the data.

In [16]:
f = '../data/credit_approval.csv'

# How do you read in that file?
df = pd.read_csv(f)

# Can you grab just the column names?
df.columns

Index([u'A', u'B', u'C', u'D', u'E', u'F', u'G', u'H', u'I', u'J', u'K', u'L', u'M', u'N', u'O', u'P'], dtype='object')

In [17]:
# How many rows and columns does the dataframe have?
df.shape

(690, 16)

In [18]:
# Now, look at the first 5 lines
df.head()

A      B      C  D  E  F  G     H  I  J  K  L  M    N    O  P
0  b  30.83  0.000  u  g  w  v  1.25  t  t  1  f  g  202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t  6  f  g   43  560  +
2  a   24.5  0.500  u  g  q  h  1.50  t  f  0  f  g  280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t  5  t  g  100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f  0  f  s  120    0  +

In [19]:
# Now, look at the last 5 lines
df.tail()

A      B       C  D  E   F   G     H  I  J  K  L  M    N    O  P
685  b  21.08  10.085  y  p   e   h  1.25  f  f  0  f  g  260    0  -
686  a  22.67   0.750  u  g   c   v  2.00  f  t  2  t  g  200  394  -
687  a  25.25  13.500  y  p  ff  ff  2.00  f  t  1  t  g  200    1  -
688  b  17.92   0.205  u  g  aa   v  0.04  f  f  0  f  g  280  750  -
689  b     35   3.375  u  g   c   h  8.29  f  f  0  t  g    0    0  -

In [20]:
# Can you describe() the data? (Notice how Pandas only "describes" the numerical data!)
df.describe()

C           H          K              O
count  690.000000  690.000000  690.00000     690.000000
mean     4.758725    2.223406    2.40000    1017.385507
std      4.978163    3.346513    4.86294    5210.102598
min      0.000000    0.000000    0.00000       0.000000
25%      1.000000    0.165000    0.00000       0.000000
50%      2.750000    1.000000    0.00000       5.000000
75%      7.207500    2.625000    3.00000     395.500000
max     28.000000   28.500000   67.00000  100000.000000

In [23]:
# Let's sort on column H
df.sort('H')

A      B       C  D   E   F   G       H  I  J   K  L  M    N       O  P
351  b  22.17   0.585  y   p  ff  ff   0.000  f  f   0  f  g  100       0  -
633  b  32.42   2.165  y   p   k  ff   0.000  f  f   0  f  g  120       0  -
456  b  34.58   0.000  ?   ?   ?   ?   0.000  f  f   0  f  p    ?       0  -
585  b  73.42  17.750  u   g  ff  ff   0.000  t  f   0  t  g    0       0  +
584  a  28.08  15.000  y   p   e   z   0.000  t  f   0  f  g    0   13212  +
345  b  62.75   7.000  u   g   e   z   0.000  f  f   0  f  g    0      12  -
53   b  34.92   2.500  u   g   w   v   0.000  t  f   0  t  g  239     200  +
261  a  52.17   0.000  y   p  ff  ff   0.000  f  f   0  f  g    0       0  -
350  a  26.17   2.000  u   g   j   j   0.000  f  f   0  t  g  276       1  -
448  b  31.25   1.125  u   g  ff  ff   0.000  f  t   1  f  g   96      19  -
447  a  27.33   1.665  u   g  ff  ff   0.000  f  f   0  f  g  340       1  -
419  a  26.58   2.540  y   p  ff  ff   0.000  f  f   0  t  g  180      60  -
305  b  48.75  26.335  y   p  ff  ff   0.000  t  f   0  t  g    0       0  -
445  a      ?  11.250  u   g  ff  ff   0.000  f  f   0  f  g    ?    5200  -
444  a  17.67   0.000  y   p   j  ff   0.000  f  f   0  f  g   86       0  -
413  b  40.58   1.500  u   g   i  bb   0.000  f  f   0  f  s  300       0  -
632  a  38.75   1.500  u   g  ff  ff   0.000  f  f   0  f  g   76       0  -
592  b  23.17   0.000  ?   ?   ?   ?   0.000  f  f   0  f  p    ?       0  +
309  a  31.75   3.000  y   p   j   j   0.000  f  f   0  f  g  160      20  -
343  b  33.75   2.750  u   g   i  bb   0.000  f  f   0  f  g  180       0  -
89   a     49   1.500  u   g   j   j   0.000  t  f   0  t  g  100      27  -
80   b  49.58  19.000  u   g  ff  ff   0.000  t  t   1  f  g   94       0  -
464  a     23   1.835  u   g   j   j   0.000  f  t   1  f  g  200      53  -
263  a  20.42  10.500  y   p   x   h   0.000  f  f   0  t  g  154      32  -
463  a  36.58   0.290  u   g  ff  ff   0.000  f  t  10  f  g  200      18  -
264  b  50.75   0.585  u   g  ff  ff   0.000  f  f   0  f  g  145       0  -
318  b  19.17   0.000  y   p   m  bb   0.000  f  f   0  t  s  500       1  +
649  a  35.17   3.750  u   g  ff  ff   0.000  f  t   6  f  g    0     200  -
601  ?  42.25   1.750  y   p   ?   ?   0.000  f  f   0  t  g  150       1  -
317  b   17.5  22.000  l  gg  ff   o   0.000  f  f   0  t  p  450  100000  +
..  ..    ...     ... ..  ..  ..  ..     ... .. ..  .. .. ..  ...     ... ..
108  b  40.92   2.250  y   p   x   h  10.000  t  f   0  t  g  176       0  -
210  b  39.33   5.875  u   g  cc   h  10.000  t  t  14  t  g  399       0  +
321  a  18.08   0.375  l  gg  cc  ff  10.000  f  f   0  t  s  300       0  +
477  b  39.17   2.500  y   p   i   h  10.000  f  f   0  t  s  200       0  -
73   b  44.25   0.500  u   g   m   v  10.750  t  f   0  f  s  400       0  -
164  a  60.58  16.500  u   g   q   v  11.000  t  f   0  t  g   21   10561  +
242  b  28.33   5.000  u   g   w   v  11.000  t  f   0  t  g   70       0  +
116  b  37.75   7.000  u   g   q   h  11.500  t  t   7  t  g  300       5  -
71   b  34.83   4.000  u   g   d  bb  12.500  t  f   0  t  g    ?       0  -
185  b  48.75   8.500  u   g   c   h  12.500  t  t   9  f  g  181    1655  +
550  b  76.75  22.290  u   g   e   z  12.750  t  t   1  t  g    0     109  +
564  b  42.17   5.040  u   g   q   h  12.750  t  f   0  t  g   92       0  +
130  b  67.75   5.500  u   g   e   z  13.000  t  t   1  t  g    0       0  +
72   a  38.58   5.000  u   g  cc   v  13.500  t  f   0  t  g  980       0  -
365  b  42.83   1.250  u   g   m   v  13.875  f  t   1  t  g  352     112  -
231  a  47.42   3.000  u   g   x   v  13.875  t  t   2  t  g  519    1704  +
246  b     45   8.500  u   g  cc   h  14.000  t  t   1  t  g   88    2000  +
118  b  57.83   7.040  u   g   m   v  14.000  t  t   6  t  g  360    1332  +
205  a  35.42  12.000  u   g   q   h  14.000  t  t   8  f  g    0    6590  +
39   b  54.58   9.415  u   g  ff  ff  14.415  t  t  11  t  g   30     300  +
1

## Working with dataframes

Pandas has a ton of functionality for manipulating and wrangling the data. Let's look at a bunch of different ways to select and subset our data.

### Selecting columns and rows

There are multiple ways to select by both rows and columns. From index to slicing to label to position, there are a variety of methods to suit your data wrangling needs.

Let's select just the mpg column from the `auto_mpg DataFrame`. This works similar to how you would access values from a dictionary:

In [24]:
auto_mpg['mpg']

0      18
1      15
2      18
3      16
4      17
5      15
6      14
7      14
8      14
9      15
10     15
11     14
12     15
13     14
14     24
15     22
16     18
17     21
18     27
19     26
20     25
21     24
22     25
23     26
24     21
25     10
26     10
27     11
28      9
29     27
       ..
368    27
369    34
370    31
371    29
372    27
373    24
374    23
375    36
376    37
377    31
378    38
379    36
380    36
381    36
382    34
383    38
384    32
385    38
386    25
387    38
388    26
389    22
390    32
391    36
392    27
393    27
394    44
395    32
396    28
397    31
Name: mpg, dtype: float64

You can do exactly the same thing by using mpg as an attribute:

In [25]:
auto_mpg.mpg

0      18
1      15
2      18
3      16
4      17
5      15
6      14
7      14
8      14
9      15
10     15
11     14
12     15
13     14
14     24
15     22
16     18
17     21
18     27
19     26
20     25
21     24
22     25
23     26
24     21
25     10
26     10
27     11
28      9
29     27
       ..
368    27
369    34
370    31
371    29
372    27
373    24
374    23
375    36
376    37
377    31
378    38
379    36
380    36
381    36
382    34
383    38
384    32
385    38
386    25
387    38
388    26
389    22
390    32
391    36
392    27
393    27
394    44
395    32
396    28
397    31
Name: mpg, dtype: float64

To extract rows from a `DataFrame`, you can use the slice method, similar to how you would slice a list. Here's how we would grab rows 7-13 from the wine `DataFrame`:

In [26]:
wine[7:14]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
7   14.06        2.15  2.61        17.6        121           2.60        2.51   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
10  14.10        2.16  2.30        18.0        105           2.95        3.32   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
12  13.75        1.73  2.41        16.0         89           2.60        2.76   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
7                   0.31             1.25   5.05  1.06      3.58     1295   
8                   0.29             1.98   5.20  1.08      2.85     1045   
9                   0.22             1.85   7.22  1.01      3.55     1045   
10                  0.22             2.38   5.75  1.25      3.17     1510   
11                  0.26             1.57   5.00  1.17      2.82     1280   
12                  0.29             1.81   5.60  1.15      2.90     1320   
13                  0.43             2.81   5.40  1.25      2.73     1150   

    wine_type  
7           1  
8           1  
9           1  
10          1  
11          1  
12          1  
13          1

Pandas also has tools for purely label-based selection of rows and columns using the `.loc` indexer. The `.loc` indexer takes input as `[row, column]`. 

For example, let's say we wanted to select the abv value in the 8th instance in our wine `DataFrame`:

In [27]:
wine.loc[8,'abv']

14.83

We can also use `.loc` to grab slices. It's important to note that `.loc` interprets the index as a *label*. This means that, if we select a range, it will grab the last item in the range, unlike slicing in a list. The index is the label for the rows. Let's grab the abv for rows 8 to 11 from the wine `DataFrame`.

In [28]:
wine.loc[8:11, 'abv']

8     14.83
9     13.86
10    14.10
11    14.12
Name: abv, dtype: float64

And, as you might expect, we can select multiple columns by passing in a list of column names. Let's also grab ash and color for rows 8 to 11.

In [29]:
wine.loc[8:11, ['abv', 'ash', 'color']]

abv   ash  color
8   14.83  2.17   5.20
9   13.86  2.27   7.22
10  14.10  2.30   5.75
11  14.12  2.32   5.00

Finally, let's just grab all columns for rows 8 to 11.

In [30]:
wine.loc[8:11, :]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
10  14.10        2.16  2.30        18.0        105           2.95        3.32   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
8                   0.29             1.98   5.20  1.08      2.85     1045   
9                   0.22             1.85   7.22  1.01      3.55     1045   
10                  0.22             2.38   5.75  1.25      3.17     1510   
11                  0.26             1.57   5.00  1.17      2.82     1280   

    wine_type  
8           1  
9           1  
10          1  
11          1

So, `.loc` provides functionality for a very specific and precise selection method.

Pandas has tools for purely position-based selection of rows and columns using the `.iloc` indexer, which works exactly how slicing a list works. The `.iloc` indexer also takes input as `[row, column]`, but takes only integer input. If we wanted to access the 60th row and the model value from `auto_mpg`, it would look like this (remember that integer indexing is 0-based):

In [31]:
auto_mpg.iloc[60, 6]

72

To grab rows 60-63 and the last three columns from the `auto_mpg DataFrame`, we would need to do the following:

In [32]:
auto_mpg.iloc[60:64, 6:9]

model  origin             car_name
60     72       1       chevrolet vega
61     72       1  ford pinto runabout
62     72       1     chevrolet impala
63     72       1     pontiac catalina

`.iloc` again works like slicing a list, based on position, so it does not grab the last item, like `.loc` does.

To grab all values and those last three columns from the `auto_mpg DataFrame`:

In [33]:
auto_mpg.iloc[:, 6:9]

model  origin                           car_name
0       70       1          chevrolet chevelle malibu
1       70       1                  buick skylark 320
2       70       1                 plymouth satellite
3       70       1                      amc rebel sst
4       70       1                        ford torino
5       70       1                   ford galaxie 500
6       70       1                   chevrolet impala
7       70       1                  plymouth fury iii
8       70       1                   pontiac catalina
9       70       1                 amc ambassador dpl
10      70       1                dodge challenger se
11      70       1                 plymouth 'cuda 340
12      70       1              chevrolet monte carlo
13      70       1            buick estate wagon (sw)
14      70       3              toyota corona mark ii
15      70       1                    plymouth duster
16      70       1                         amc hornet
17      70       1                      ford maverick
18      70       3                       datsun pl510
19      70       2       volkswagen 1131 deluxe sedan
20      70       2                        peugeot 504
21      70       2                        audi 100 ls
22      70       2                           saab 99e
23      70       2                           bmw 2002
24      70       1                        amc gremlin
25      70       1                          ford f250
26      70       1                          chevy c20
27      70       1                         dodge d200
28      70       1                           hi 1200d
29      71       3                       datsun pl510
..     ...     ...                                ...
368     82       1           chevrolet cavalier wagon
369     82       1          chevrolet cavalier 2-door
370     82       1         pontiac j2000 se hatchback
371     82       1                     dodge aries se
372     82       1                    pontiac phoenix
373     82       1               ford fairmont futura
374     82       1                     amc concord dl
375     82       2                volkswagen rabbit l
376     82       3                 mazda glc custom l
377     82       3                   mazda glc custom
378     82       1             plymouth horizon miser
379     82       1                     mercury lynx l
380     82       3                   nissan stanza xe
381     82       3                       honda accord
382     82       3                     toyota corolla
383     82       3                        honda civic
384     82       3                 honda civic (auto)
385     82       3                      datsun 310 gx
386     82       1              buick century limited
387     82       1  oldsmobile cutlass ciera (diesel)
388     82       1         chrysler lebaron medallion
389     82       1                     ford granada l
390     82       3                   toyota celica gt
391     82       1                  dodge charger 2.2
392     82       1                   chevrolet camaro
393     82       1                    ford mustang gl
394     82       2                          vw pickup
395     82       1                      dodge rampage
396     82       1                        ford ranger
397     82       1                         chevy s-10

[398 rows x 3 columns]

One of my favorite methods for selecting data is through boolean indexing. Boolean indexing is similar to the WHERE clause in SQL in that it allows you to filter out data based on certain criteria. Let's see how this works.

Let's select from the wine `DataFrame` where `wine_type` is type 1.

In [34]:
wine[wine['wine_type'] == 1]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
0   14.23        1.71  2.43        15.6        127           2.80        3.06   
1   13.20        1.78  2.14        11.2        100           2.65        2.76   
2   13.16        2.36  2.67        18.6        101           2.80        3.24   
3   14.37        1.95  2.50        16.8        113           3.85        3.49   
4   13.24        2.59  2.87        21.0        118           2.80        2.69   
5   14.20        1.76  2.45        15.2        112           3.27        3.39   
6   14.39        1.87  2.45        14.6         96           2.50        2.52   
7   14.06        2.15  2.61        17.6        121           2.60        2.51   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
10  14.10        2.16  2.30        18.0        105           2.95        3.32   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
12  13.75        1.73  2.41        16.0         89           2.60        2.76   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   
14  14.38        1.87  2.38        12.0        102           3.30        3.64   
15  13.63        1.81  2.70        17.2        112           2.85        2.91   
16  14.30        1.92  2.72        20.0        120           2.80        3.14   
17  13.83        1.57  2.62        20.0        115           2.95        3.40   
18  14.19        1.59  2.48        16.5        108           3.30        3.93   
19  13.64        3.10  2.56        15.2        116           2.70        3.03   
20  14.06        1.63  2.28        16.0        126           3.00        3.17   
21  12.93        3.80  2.65        18.6        102           2.41        2.41   
22  13.71        1.86  2.36        16.6        101           2.61        2.88   
23  12.85        1.60  2.52        17.8         95           2.48        2.37   
24  13.50        1.81  2.61        20.0         96           2.53        2.61   
25  13.05        2.05  3.22        25.0        124           2.63        2.68   
26  13.39        1.77  2.62        16.1         93           2.85        2.94   
27  13.30        1.72  2.14        17.0         94           2.40        2.19   
28  13.87        1.90  2.80        19.4        107           2.95        2.97   
29  14.02        1.68  2.21        16.0         96           2.65        2.33   
30  13.73        1.50  2.70        22.5        101           3.00        3.25   
31  13.58        1.66  2.36        19.1        106           2.86        3.19   
32  13.68        1.83  2.36        17.2        104           2.42        2.69   
33  13.76        1.53  2.70        19.5        132           2.95        2.74   
34  13.51        1.80  2.65        19.0        110           2.35        2.53   
35  13.48        1.81  2.41        20.5        100           2.70        2.98   
36  13.28        1.64  2.84        15.5        110           2.60        2.68   
37  13.05        1.65  2.55        18.0         98           2.45        2.43   
38  13.07        1.50  2.10        15.5         98           2.40        2.64   
39  14.22        3.99  2.51        13.2        128           3.00        3.04   
40  13.56        1.71  2.31        16.2        117           3.15        3.29   
41  13.41        3.84  2.12        18.8         90           2.45        2.68   
42  13.88        1.89  2.59        15.0        101           3.25        3.56   
43  13.24        3.98  2.29        17.5        103           2.64        2.63   
44  13.05        1.77  2.10        17.0        107           3.00        3.00   
45  14.21        4.04  2.44        18.9        111           2.85        2.65   
46  14.38        3.59  2.28        16.0        102           3.25        3.17   
47  13.90        1.68  2.12        16.0        101           3.10        3.39   
48  14.10        2.02  2.40        18

This works with any comparison operators, like >, < >=, !=, and so on. For example, we can select everything from the wine `DataFrame` where the value in the magnesium column is less than 100.

In [35]:
wine[wine['magnesium'] < 100]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  \
6    14.39        1.87  2.45        14.6         96           2.50   
8    14.83        1.64  2.17        14.0         97           2.80   
9    13.86        1.35  2.27        16.0         98           2.98   
11   14.12        1.48  2.32        16.8         95           2.20   
12   13.75        1.73  2.41        16.0         89           2.60   
13   14.75        1.73  2.39        11.4         91           3.10   
23   12.85        1.60  2.52        17.8         95           2.48   
24   13.50        1.81  2.61        20.0         96           2.53   
26   13.39        1.77  2.62        16.1         93           2.85   
27   13.30        1.72  2.14        17.0         94           2.40   
29   14.02        1.68  2.21        16.0         96           2.65   
37   13.05        1.65  2.55        18.0         98           2.45   
38   13.07        1.50  2.10        15.5         98           2.40   
41   13.41        3.84  2.12        18.8         90           2.45   
50   13.05        1.73  2.04        12.4         92           2.72   
51   13.83        1.65  2.60        17.2         94           2.45   
59   12.37        0.94  1.36        10.6         88           1.98   
62   13.67        1.25  1.92        18.0         94           2.10   
63   12.37        1.13  2.16        19.0         87           3.50   
65   12.37        1.21  2.56        18.1         98           2.42   
66   13.11        1.01  1.70        15.0         78           2.98   
67   12.37        1.17  1.92        19.6         78           2.11   
71   13.86        1.51  2.67        25.0         86           2.95   
72   13.49        1.66  2.24        24.0         87           1.88   
75   11.66        1.88  1.92        16.0         97           1.61   
76   13.03        0.90  1.71        16.0         86           1.95   
80   12.00        0.92  2.00        19.0         86           2.42   
81   12.72        1.81  2.20        18.8         86           2.20   
82   12.08        1.13  2.51        24.0         78           2.00   
83   13.05        3.86  2.32        22.5         85           1.65   
..     ...         ...   ...         ...        ...            ...   
128  12.37        1.63  2.30        24.5         88           2.22   
129  12.04        4.30  2.38        22.0         80           2.10   
132  12.81        2.31  2.40        24.0         98           1.15   
134  12.51        1.24  2.25        17.5         85           2.00   
135  12.60        2.46  2.20        18.5         94           1.62   
136  12.25        4.72  2.54        21.0         89           1.38   
137  12.53        5.51  2.64        25.0         96           1.79   
138  13.49        3.59  2.19        19.5         88           1.62   
140  12.93        2.81  2.70        21.0         96           1.54   
141  13.36        2.56  2.35        20.0         89           1.40   
142  13.52        3.17  2.72        23.5         97           1.55   
143  13.62        4.95  2.35        20.0         92           2.00   
146  13.88        5.04  2.23        20.0         80           0.98   
147  12.87        4.61  2.48        21.5         86           1.70   
148  13.32        3.24  2.38        21.5         92           1.93   
153  13.23        3.30  2.28        18.5         98           1.80   
155  13.17        5.19  2.32        22.0         93           1.74   
156  13.84        4.12  2.38        19.5         89           1.80   
157  12.45        3.03  2.64        27.0         97           1.90   
158  14.34        1.68  2.70        25.0         98           2.80   
159  13.48        1.67  2.64        22.5         89           2.60   
160  12.36        3.83  2.38        21.0         88           2.30   
164  13.78        2.76  2.30        22.0         90           1.35   
165  13.73        4.36  2.26        22.5         88           1.28   
167  12.82        3.37  2.30        19.5         88           1.48   
170  12.20       

You can also say 'not' with the tilde: ~

Let's select from the wine `DataFrame` where magnesium is NOT less than 100, which is equivalent to saying greater than or equal to.

In [36]:
wine[~(wine['magnesium'] < 100)]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  \
0    14.23        1.71  2.43        15.6        127           2.80   
1    13.20        1.78  2.14        11.2        100           2.65   
2    13.16        2.36  2.67        18.6        101           2.80   
3    14.37        1.95  2.50        16.8        113           3.85   
4    13.24        2.59  2.87        21.0        118           2.80   
5    14.20        1.76  2.45        15.2        112           3.27   
7    14.06        2.15  2.61        17.6        121           2.60   
10   14.10        2.16  2.30        18.0        105           2.95   
14   14.38        1.87  2.38        12.0        102           3.30   
15   13.63        1.81  2.70        17.2        112           2.85   
16   14.30        1.92  2.72        20.0        120           2.80   
17   13.83        1.57  2.62        20.0        115           2.95   
18   14.19        1.59  2.48        16.5        108           3.30   
19   13.64        3.10  2.56        15.2        116           2.70   
20   14.06        1.63  2.28        16.0        126           3.00   
21   12.93        3.80  2.65        18.6        102           2.41   
22   13.71        1.86  2.36        16.6        101           2.61   
25   13.05        2.05  3.22        25.0        124           2.63   
28   13.87        1.90  2.80        19.4        107           2.95   
30   13.73        1.50  2.70        22.5        101           3.00   
31   13.58        1.66  2.36        19.1        106           2.86   
32   13.68        1.83  2.36        17.2        104           2.42   
33   13.76        1.53  2.70        19.5        132           2.95   
34   13.51        1.80  2.65        19.0        110           2.35   
35   13.48        1.81  2.41        20.5        100           2.70   
36   13.28        1.64  2.84        15.5        110           2.60   
39   14.22        3.99  2.51        13.2        128           3.00   
40   13.56        1.71  2.31        16.2        117           3.15   
42   13.88        1.89  2.59        15.0        101           3.25   
43   13.24        3.98  2.29        17.5        103           2.64   
..     ...         ...   ...         ...        ...            ...   
77   11.84        2.89  2.23        18.0        112           1.72   
78   12.33        0.99  1.95        14.8        136           1.90   
79   12.70        3.87  2.40        23.0        101           2.83   
95   12.47        1.52  2.20        19.0        162           2.50   
96   11.81        2.12  2.74        21.5        134           1.60   
110  11.46        3.74  1.82        19.5        107           3.18   
112  11.76        2.68  2.92        20.0        103           1.75   
117  12.42        1.61  2.19        22.5        108           2.00   
121  11.56        2.05  3.23        28.5        119           3.18   
122  12.42        4.43  2.73        26.5        102           2.20   
130  12.86        1.35  2.32        18.0        122           1.51   
131  12.88        2.99  2.40        20.0        104           1.30   
133  12.70        3.55  2.36        21.5        106           1.70   
139  12.84        2.96  2.61        24.0        101           2.32   
144  12.25        3.88  2.20        18.5        112           1.38   
145  13.16        3.57  2.15        21.0        102           1.50   
149  13.08        3.90  2.36        21.5        113           1.41   
150  13.50        3.12  2.62        24.0        123           1.40   
151  12.79        2.67  2.48        22.0        112           1.48   
152  13.11        1.90  2.75        25.5        116           2.20   
154  12.58        1.29  2.10        20.0        103           1.48   
161  13.69        3.26  2.54        20.0        107           1.83   
162  12.85        3.27  2.58        22.0        106           1.65   
163  12.96        3.45  2.35        18.5        106           1.39   
166  13.45        3.70  2.60        23.0        111           1.70   
168  13.58       

It's also possible to combine these boolean indexers. Make sure you enclose them in parentheses. This is something I usually forget.

Let's select from wine where magnesium is less than 100 and the type of wine is type 1.

In [37]:
wine[(wine['magnesium'] < 100) & (wine['wine_type'] == 1)]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
6   14.39        1.87  2.45        14.6         96           2.50        2.52   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
9   13.86        1.35  2.27        16.0         98           2.98        3.15   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
12  13.75        1.73  2.41        16.0         89           2.60        2.76   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   
23  12.85        1.60  2.52        17.8         95           2.48        2.37   
24  13.50        1.81  2.61        20.0         96           2.53        2.61   
26  13.39        1.77  2.62        16.1         93           2.85        2.94   
27  13.30        1.72  2.14        17.0         94           2.40        2.19   
29  14.02        1.68  2.21        16.0         96           2.65        2.33   
37  13.05        1.65  2.55        18.0         98           2.45        2.43   
38  13.07        1.50  2.10        15.5         98           2.40        2.64   
41  13.41        3.84  2.12        18.8         90           2.45        2.68   
50  13.05        1.73  2.04        12.4         92           2.72        3.27   
51  13.83        1.65  2.60        17.2         94           2.45        2.99   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
6                   0.30             1.98   5.25  1.02      3.58     1290   
8                   0.29             1.98   5.20  1.08      2.85     1045   
9                   0.22             1.85   7.22  1.01      3.55     1045   
11                  0.26             1.57   5.00  1.17      2.82     1280   
12                  0.29             1.81   5.60  1.15      2.90     1320   
13                  0.43             2.81   5.40  1.25      2.73     1150   
23                  0.26             1.46   3.93  1.09      3.63     1015   
24                  0.28             1.66   3.52  1.12      3.82      845   
26                  0.34             1.45   4.80  0.92      3.22     1195   
27                  0.27             1.35   3.95  1.02      2.77     1285   
29                  0.26             1.98   4.70  1.04      3.59     1035   
37                  0.29             1.44   4.25  1.12      2.51     1105   
38                  0.28             1.37   3.70  1.18      2.69     1020   
41                  0.27             1.48   4.28  0.91      3.00     1035   
50                  0.17             2.91   7.20  1.12      2.91     1150   
51                  0.22             2.29   5.60  1.24      3.37     1265   

    wine_type  
6           1  
8           1  
9           1  
11          1  
12          1  
13          1  
23          1  
24          1  
26          1  
27          1  
29          1  
37          1  
38          1  
41          1  
50          1  
51          1

If you wanted to, you could just keep on chaining the booleans together. Let's add on where the abv is greater than 14.

In [38]:
wine[(wine['magnesium'] < 100) & (wine['wine_type'] == 1) & (wine['abv'] > 14)]

abv  malic_acid   ash  alcalinity  magnesium  total_phenols  flavanoids  \
6   14.39        1.87  2.45        14.6         96           2.50        2.52   
8   14.83        1.64  2.17        14.0         97           2.80        2.98   
11  14.12        1.48  2.32        16.8         95           2.20        2.43   
13  14.75        1.73  2.39        11.4         91           3.10        3.69   
29  14.02        1.68  2.21        16.0         96           2.65        2.33   

    nonflavanoid_phenols  proanthocyacins  color   hue  dilution  proline  \
6                   0.30             1.98   5.25  1.02      3.58     1290   
8                   0.29             1.98   5.20  1.08      2.85     1045   
11                  0.26             1.57   5.00  1.17      2.82     1280   
13                  0.43             2.81   5.40  1.25      2.73     1150   
29                  0.26             1.98   4.70  1.04      3.59     1035   

    wine_type  
6           1  
8           1  
11          1  
13          1  
29          1

Another method of selecting data is using the `isin()` function. If you pass in a list to `isin()`, it will return a `DataFrame` of booleans. True means that the value at that index is in the list you passed into `isin()`.

Let's take the first five rows of the `auto_mpg DataFrame` and check for certain values existing in the `DataFrame`.

In [39]:
auto_mpg_5 = auto_mpg.head()

vals = [8, 150, 12.0, 'ford torino']
auto_mpg_5.isin(vals)

mpg cylinders displacement horsepower weight acceleration  model origin  \
0  False      True        False      False  False         True  False  False   
1  False      True        False      False  False        False  False  False   
2  False      True        False      False  False        False  False  False   
3  False      True        False      False  False         True  False  False   
4  False      True        False      False  False        False  False  False   

  car_name  
0    False  
1    False  
2    False  
3    False  
4     True

If it says `True`, it means that one of the values from the `vals` list occurs there.

## Lesson: let's try some of these on some data!

In [40]:
# Extract column C from the credit_approval dataframe we read in above
df['C']

0       0.000
1       4.460
2       0.500
3       1.540
4       5.625
5       4.000
6       1.040
7      11.585
8       0.500
9       4.915
10      0.830
11      1.835
12      6.000
13      6.040
14     10.500
15      4.415
16      0.875
17      5.875
18      0.250
19      8.585
20     11.250
21      1.000
22      8.000
23     14.500
24      6.500
25      0.585
26     13.000
27     18.500
28      8.500
29      1.040
        ...  
660     9.000
661     3.500
662     1.500
663     4.000
664     1.500
665     0.040
666    11.750
667     0.540
668     0.500
669     2.040
670     5.835
671    12.835
672     0.835
673     2.000
674     2.500
675     1.040
676    10.665
677     7.250
678    10.210
679     1.250
680     0.290
681     1.000
682     3.290
683     0.750
684     3.290
685    10.085
686     0.750
687    13.500
688     0.205
689     3.375
Name: C, dtype: float64

In [44]:
# Slice rows 5-10 from the credit_approval dataframe
df[5:11]

A      B       C  D  E   F  G      H  I  J  K  L  M    N      O  P
5   b  32.08   4.000  u  g   m  v  2.500  t  f  0  t  g  360      0  +
6   b  33.17   1.040  u  g   r  h  6.500  t  f  0  t  g  164  31285  +
7   a  22.92  11.585  u  g  cc  v  0.040  t  f  0  f  g   80   1349  +
8   b  54.42   0.500  y  p   k  h  3.960  t  f  0  f  g  180    314  +
9   b   42.5   4.915  y  p   w  v  3.165  t  f  0  t  g   52   1442  +
10  b  22.08   0.830  u  g   c  h  2.165  f  f  0  t  g  128      0  +

In [45]:
# How would you look up the value for the 13th row in column C by label (loc)?
df.loc[13, 'C']

6.04

In [46]:
# How would you look up the same thing by position (iloc)?
df.iloc[13, 2]

6.04

In [47]:
# What if I wanted to select all data from credit_approval based on column C being greater than 5?
df[df['C'] > 5]

A      B       C  D  E   F   G       H  I  J   K  L  M    N      O  P
4    b  20.17   5.625  u  g   w   v   1.710  t  f   0  f  s  120      0  +
7    a  22.92  11.585  u  g  cc   v   0.040  t  f   0  f  g   80   1349  +
12   a  38.25   6.000  u  g   k   v   1.000  t  f   0  t  g    0      0  +
13   b  48.08   6.040  u  g   k   v   0.040  f  f   0  f  g    0   2690  +
14   a  45.83  10.500  u  g   q   v   5.000  t  t   7  t  g    0      0  +
17   a  23.25   5.875  u  g   q   v   3.170  t  t  10  f  g  120    245  +
19   a  19.17   8.585  u  g  cc   h   0.750  t  t   7  f  g   96      0  +
20   b     25  11.250  u  g   c   v   2.500  t  t  17  f  g  200   1208  +
22   a  47.75   8.000  u  g   c   v   7.875  t  t   6  t  g    0   1260  +
23   a  27.42  14.500  u  g   x   h   3.085  t  t   1  f  g  120     11  +
24   a  41.17   6.500  u  g   q   v   0.500  t  t   3  t  g  145      0  +
26   a     47  13.000  u  g   i  bb   5.165  t  t   9  t  g    0      0  +
27   b  56.58  18.500  u  g   d  bb  15.000  t  t  17  t  g    0      0  +
28   b  57.42   8.500  u  g   e   h   7.000  t  t   3  f  g    0      0  +
30   b  29.25  14.790  u  g  aa   v   5.040  t  t   5  t  g  168      0  +
31   b     42   9.790  u  g   x   h   7.960  t  t   8  f  g    0      0  +
32   b   49.5   7.585  u  g   i  bb   7.585  t  t  15  t  g    0   5000  +
33   a  36.75   5.125  u  g   e   v   5.000  t  f   0  t  g    0   4000  +
34   a  22.58  10.750  u  g   q   v   0.415  t  t   5  t  g    0    560  +
37   a     23  11.750  u  g   x   h   0.500  t  t   2  t  g  300    551  +
39   b  54.58   9.415  u  g  ff  ff  14.415  t  t  11  t  g   30    300  +
40   b  34.17   9.170  u  g   c   v   4.500  t  t  12  t  g    0    221  +
41   b  28.92  15.000  u  g   c   h   5.335  t  t  11  f  g    0   2283  +
43   b  39.58  13.915  u  g   w   v   8.625  t  t   6  t  g   70      0  +
44   b  56.42  28.000  y  p   c   v  28.500  t  t  40  f  g    0     15  +
45   b  54.33   6.750  u  g   c   h   2.625  t  t  11  t  g    0    284  +
55   b  23.33  11.625  y  p   w   v   0.835  t  f   0  t  g  160    300  +
60   b  56.75  12.250  u  g   m   v   1.250  t  t   4  t  g  200      0  +
61   b  31.67  16.165  u  g   d   v   3.000  t  t   9  f  g  250    730  +
68   b  19.42   6.500  u  g   w   h   1.460  t  t   7  f  g   80   2954  +
..  ..    ...     ... .. ..  ..  ..     ... .. ..  .. .. ..  ...    ... ..
579  b  39.08   6.000  u  g   m   v   1.290  t  t   5  t  g  108   1097  +
583  b  32.67   9.000  y  p   w   h   5.250  t  f   0  t  g  154      0  +
584  a  28.08  15.000  y  p   e   z   0.000  t  f   0  f  g    0  13212  +
585  b  73.42  17.750  u  g  ff  ff   0.000  t  f   0  t  g    0      0  +
586  b  64.08  20.000  u  g   x   h  17.500  t  t   9  t  g    0   1000  +
587  b  51.58  15.000  u  g   c   v   8.500  t  t   9  f  g    0      0  +
590  b  30.17   6.500  u  g  cc   v   3.125  t  t   8  f  g  330   1200  +
593  b  34.17   5.250  u  g   w   v   0.085  f  f   0  t  g  290      6  +
597  a   21.5   6.000  u  g  aa   v   2.500  t  t   3  f  g   80    918  +
613  b  19.33  10.915  u  g   c  bb   0.585  f  t   2  t  g  200      7  -
617  b  32.25  14.000  y  p  ff  ff   0.000  f  t   2  f  g  160      1  -
619  b  18.42  10.415  y  p  aa   v   0.125  t  f   0  f  g  120    375  -
625  b  23.75  12.000  u  g   c   v   2.085  f  f   0  f  s   80      0  -
626  b     22   7.835  y  p   i  bb   0.165  f  f   0  t  g    ?      0  -
628  b  29.25  13.000  u  g   d   h   0.500  f  f   0  f  g  228      0  -
637  b   19.5   9.585  u  g  aa   v   0.790  f  f   0  f  g   80    350  -
643  a   52.5   7.000  u  g  aa   h   3.000  f  f   0  f  g    0      0  -
646  a  20.83   8.500  u  g   c   v   0.165  f  f   0  f  g    0    351  -
647  b  24.08   9.000  u  g  aa   v   0.250  f  f   0  t  g    0      0  -
651  a  15.83   7.625  u  g   q   v   0.125  f  t   1  t  g    0    160  -
653  b   21.5  11.500  u  g   i   v   0.500  t  f   0  t  g  100     68  -
660  b  22.25   9.000  u  g  a

In [48]:
# What if I wanted to select data based on column C being greater than 5 and column F being equal to 'w'?
df[(df['C'] > 5) & (df['F'] == 'w')]

A      B       C  D  E  F  G       H  I  J   K  L  M    N     O  P
4    b  20.17   5.625  u  g  w  v   1.710  t  f   0  f  s  120     0  +
43   b  39.58  13.915  u  g  w  v   8.625  t  t   6  t  g   70     0  +
55   b  23.33  11.625  y  p  w  v   0.835  t  f   0  t  g  160   300  +
68   b  19.42   6.500  u  g  w  h   1.460  t  t   7  f  g   80  2954  +
112  b  24.58  12.500  u  g  w  v   0.875  t  f   0  t  g  260     0  -
138  a  18.83   9.500  u  g  w  v   1.625  t  t   6  t  g   40   600  +
143  b  22.33  11.000  u  g  w  v   2.000  t  t   1  f  g   80   278  +
197  b  48.17   7.625  u  g  w  h  15.500  t  t  12  f  g    0   790  +
233  b  27.67  13.750  u  g  w  v   5.750  t  f   0  t  g  487   500  +
241  b  48.25  25.085  u  g  w  v   1.750  t  t   3  f  g  120    14  +
285  b  17.58  10.000  u  g  w  h   0.165  f  t   1  f  g  120     1  -
291  a  18.25  10.000  u  g  w  v   1.000  f  t   1  f  g  120     1  -
455  b  36.17  18.125  u  g  w  v   0.085  f  f   0  f  g  320  3552  -
495  b  19.17   9.500  u  g  w  v   1.500  t  f   0  f  g  120  2206  +
566  b  23.08  11.500  u  g  w  h   2.125  t  t  11  t  g  290   284  +
583  b  32.67   9.000  y  p  w  h   5.250  t  f   0  t  g  154     0  +
593  b  34.17   5.250  u  g  w  v   0.085  f  f   0  t  g  290     6  +
670  b  47.17   5.835  u  g  w  v   5.500  f  f   0  f  g  465   150  -

In [49]:
# What if I wanted to look at a boolean DataFrame of where values are in ['t', 's', 100, 0] in credit_approval?
val_list = ['t', 's', 100, 0]

df.isin(val_list)

A      B      C      D      E      F      G      H      I      J  \
0    False  False   True  False  False  False  False  False   True   True   
1    False  False  False  False  False  False  False  False   True   True   
2    False  False  False  False  False  False  False  False   True  False   
3    False  False  False  False  False  False  False  False   True   True   
4    False  False  False  False  False  False  False  False   True  False   
5    False  False  False  False  False  False  False  False   True  False   
6    False  False  False  False  False  False  False  False   True  False   
7    False  False  False  False  False  False  False  False   True  False   
8    False  False  False  False  False  False  False  False   True  False   
9    False  False  False  False  False  False  False  False   True  False   
10   False  False  False  False  False  False  False  False  False  False   
11   False  False  False  False  False  False  False  False   True  False   
12   False  False  False  False  False  False  False  False   True  False   
13   False  False  False  False  False  False  False  False  False  False   
14   False  False  False  False  False  False  False  False   True   True   
15   False  False  False  False  False  False  False  False   True   True   
16   False  False  False  False  False  False  False  False   True   True   
17   False  False  False  False  False  False  False  False   True   True   
18   False  False  False  False  False  False  False  False   True  False   
19   False  False  False  False  False  False  False  False   True   True   
20   False  False  False  False  False  False  False  False   True   True   
21   False  False  False  False  False  False  False  False   True  False   
22   False  False  False  False  False  False  False  False   True   True   
23   False  False  False  False  False  False  False  False   True   True   
24   False  False  False  False  False  False  False  False   True   True   
25   False  False  False  False  False  False  False  False   True   True   
26   False  False  False  False  False  False  False  False   True   True   
27   False  False  False  False  False  False  False  False   True   True   
28   False  False  False  False  False  False  False  False   True   True   
29   False  False  False  False  False  False  False  False   True   True   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
660  False  False  False  False  False  False  False  False  False  False   
661  False  False  False  False  False  False  False  False  False  False   
662  False  False  False  False  False  False  False  False  False  False   
663  False  False  False  False  False  False  False  False  False  False   
664  False  False  False  False  False  False  False  False  False  False   
665  False  False  False  False  False  False  False  False  False  False   
666  False  False  False  False  False  False  False  False  False  False   
667  False  False  False  False  False  False  False  False  False   True   
668  False  False  False  False  False  False  False  False  False  False   
669  False  False  False  False  False  False  False  False  False  False   
670  False  False  False  False  False  False  False  False  False  False   
671  False  False  False  False  False  False  False  False  False  False   
672  False  False  False  False  False  False  False  False  False  False   
673  False  False  False  False  False  False  False  False  False  False   
674  False  False  False  False  False  False  False  False  False  False   
675  False  False  False  False  False  False  False  False  False  False   
676  False  False  False  False  False  False  False  False  False   True   
677  False  False  False  False  False  False  False  False  False   True   
678  False  False  False  False  False  False  False   True  False  False   
679  False  False  False  False  False  False  False   True  False  False   
680  

## Groupby

`groupby()` is just like SQL's 'group by' clause. What groupby does is a three-step process:

- Split the data
- Apply a function to the split groups
- Recombine the data

In the apply step, you can do things like apply a statistical function, filter out data, or transform the data.

Let's `groupby()` the wine_type in our wine `DataFrame`! Let's start with just `groupby()`, and then build it from there. This will produce a `DataFrame groupby` object.

In [50]:
wine.groupby('wine_type')

Not so interesting yet. This object has some attributes you can access. We can get lists of which rows are in which group by using the `.groups` attribute:

In [51]:
wine.groupby('wine_type').groups

{1: [0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  27,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  38,
  39,
  40,
  41,
  42,
  43,
  44,
  45,
  46,
  47,
  48,
  49,
  50,
  51,
  52,
  53,
  54,
  55,
  56,
  57,
  58],
 2: [59,
  60,
  61,
  62,
  63,
  64,
  65,
  66,
  67,
  68,
  69,
  70,
  71,
  72,
  73,
  74,
  75,
  76,
  77,
  78,
  79,
  80,
  81,
  82,
  83,
  84,
  85,
  86,
  87,
  88,
  89,
  90,
  91,
  92,
  93,
  94,
  95,
  96,
  97,
  98,
  99,
  100,
  101,
  102,
  103,
  104,
  105,
  106,
  107,
  108,
  109,
  110,
  111,
  112,
  113,
  114,
  115,
  116,
  117,
  118,
  119,
  120,
  121,
  122,
  123,
  124,
  125,
  126,
  127,
  128,
  129],
 3: [130,
  131,
  132,
  133,
  134,
  135,
  136,
  137,
  138,
  139,
  140,
  141,
  142,
  143,
  144,
  145,
  146,
  147,
  148,
  149,
  150,
  151,
  152,
  153,
  154,
  155,
  156,


The dataset was in order by `wine_type` to begin with, so that makes sense. To get just the keys, add the `.keys()` function to the end of that line.

In [52]:
wine.groupby('wine_type').groups.keys()

[1, 2, 3]

Let's group our `auto_mpg` dataset by cylinders, just for contrast.

In [53]:
auto_mpg.groupby('cylinders').groups

{3: [71, 111, 243, 334],
 4: [14,
  18,
  19,
  20,
  21,
  22,
  23,
  29,
  30,
  31,
  32,
  46,
  49,
  50,
  51,
  52,
  53,
  54,
  55,
  56,
  57,
  58,
  59,
  60,
  61,
  76,
  77,
  78,
  79,
  80,
  81,
  82,
  83,
  84,
  102,
  108,
  109,
  110,
  112,
  114,
  117,
  118,
  119,
  120,
  122,
  129,
  130,
  131,
  132,
  141,
  142,
  143,
  144,
  145,
  146,
  147,
  148,
  149,
  150,
  151,
  167,
  168,
  170,
  171,
  172,
  173,
  175,
  177,
  178,
  179,
  180,
  181,
  182,
  183,
  184,
  185,
  186,
  195,
  196,
  197,
  198,
  203,
  204,
  205,
  206,
  207,
  209,
  216,
  217,
  218,
  219,
  220,
  233,
  234,
  235,
  236,
  237,
  238,
  239,
  240,
  242,
  244,
  245,
  246,
  247,
  248,
  255,
  266,
  267,
  268,
  269,
  270,
  271,
  272,
  273,
  276,
  278,
  279,
  282,
  293,
  294,
  295,
  296,
  299,
  301,
  302,
  303,
  304,
  305,
  308,
  309,
  310,
  311,
  312,
  313,
  314,
  315,
  317,
  318,
  319,
  320,
  321,
  322,
  323

You can see we have four observations with three cylinders, many more with four, and so on.

Going back to the wine example, let's apply an aggregate function. Let's generate the mean of all the other values and group them by `wine_class`.

In [54]:
wine.groupby('wine_type').mean()

abv  malic_acid       ash  alcalinity   magnesium  \
wine_type                                                            
1          13.744746    2.010678  2.455593   17.037288  106.338983   
2          12.278732    1.932676  2.244789   20.238028   94.549296   
3          13.153750    3.333750  2.437083   21.416667   99.312500   

           total_phenols  flavanoids  nonflavanoid_phenols  proanthocyacins  \
wine_type                                                                     
1               2.840169    2.982373              0.290000         1.899322   
2               2.258873    2.080845              0.363662         1.630282   
3               1.678750    0.781458              0.447500         1.153542   

              color       hue  dilution      proline  
wine_type                                             
1          5.528305  1.062034  3.157797  1115.711864  
2          3.086620  1.056282  2.785352   519.507042  
3          7.396250  0.682708  1.683542   629.895833

So, the mean `abv` for wine with type 1 is 13.74, type 2 is 12.27, type 3 is 13.15. The mean `malic_acid` for wine with type 1 is 2.01, and so on. So, with one line of code, we're able to apply a function to the entire dataset and see what's going on within different groups.

Selecting from a `groupby DataFrame` works the same way as selecting from any other `DataFrame`. Let's select the abv where `wine_type` is 2.

In [60]:
wine_type_mean = wine.groupby('wine_type').mean()

wine_type_mean.loc[2, 'abv']

12.278732394366198

It's also possible to apply multiple functions to the entire `DataFrame` using the `agg()` function. Let's get not only the mean, but the count and the standard deviation as well for each value in the `DataFrame`, still grouping by `wine_type`.

In [61]:
wine.groupby('wine_type').agg(['mean', 'count', 'std'])

abv                 malic_acid                       ash  \
                mean count       std       mean count       std      mean   
wine_type                                                                   
1          13.744746    59  0.462125   2.010678    59  0.688549  2.455593   
2          12.278732    71  0.537964   1.932676    71  1.015569  2.244789   
3          13.153750    48  0.530241   3.333750    48  1.087906  2.437083   

                          alcalinity     ...         color       hue        \
          count       std       mean     ...           std      mean count   
wine_type                                ...                                 
1            59  0.227166  17.037288     ...      1.238573  1.062034    59   
2            71  0.315467  20.238028     ...      0.924929  1.056282    71   
3            48  0.184690  21.416667     ...      2.310942  0.682708    48   

                     dilution                      proline                    
                std      mean count       std         mean count         std  
wine_type                                                                     
1          0.116483  3.157797    59  0.357077  1115.711864    59  221.520767  
2          0.202937  2.785352    71  0.496573   519.507042    71  157.211220  
3          0.114441  1.683542    48  0.272111   629.895833    48  115.097043  

[3 rows x 39 columns]

It's also possible to run different functions on different columns. Let's get the mean for abv, the standard deviation for ash, and the sum of the values for hue. To do this, you'll need to create a dictionary with these functions, with the column names as the dictionary keys.

In [62]:
multiple_funcs = {'abv': 'std', 'ash': 'mean', 'hue': sum}

wine.groupby('wine_type').agg(multiple_funcs)

hue       ash       abv
wine_type                            
1          62.660  2.455593  0.462125
2          74.996  2.244789  0.537964
3          32.770  2.437083  0.530241

## Lesson: Groupby galore

Let's take this one step at a time.

In [64]:
# Let's group credit_approval by column G.
grouped = df.groupby('G')

In [65]:
# Can you generate a list of all of the groups in the groupby object we just made?
grouped.groups

{'?': [206, 270, 330, 456, 479, 539, 592, 601, 622],
 'bb': [26,
  27,
  32,
  48,
  70,
  71,
  85,
  95,
  97,
  100,
  113,
  126,
  128,
  132,
  133,
  148,
  159,
  171,
  180,
  181,
  186,
  194,
  195,
  201,
  209,
  225,
  234,
  236,
  248,
  318,
  327,
  334,
  342,
  343,
  347,
  378,
  379,
  380,
  382,
  388,
  391,
  392,
  402,
  404,
  413,
  450,
  458,
  527,
  532,
  538,
  541,
  544,
  551,
  595,
  609,
  613,
  626,
  640,
  650],
 'dd': [266, 364, 371, 375, 513, 521],
 'ff': [39,
  80,
  111,
  141,
  212,
  217,
  261,
  264,
  278,
  286,
  294,
  295,
  296,
  300,
  305,
  306,
  307,
  321,
  351,
  359,
  361,
  362,
  383,
  397,
  405,
  410,
  415,
  419,
  428,
  430,
  435,
  436,
  437,
  438,
  441,
  444,
  445,
  447,
  448,
  463,
  485,
  488,
  502,
  533,
  535,
  536,
  573,
  585,
  617,
  627,
  632,
  633,
  649,
  655,
  669,
  678,
  687],
 'h': [1,
  2,
  6,
  8,
  10,
  11,
  18,
  19,
  23,
  25,
  28,
  31,
  36,
  37,
  41,
  

In [73]:
# Let's use mean() on credit_approval_group to get the mean of our numeric values.
grouped.agg('mean')
# can also do grouped.mean()

C         H         K             O
G                                              
?    1.106667  0.069444  0.000000     37.888889
bb   4.699068  3.034746  2.915254   1762.271186
dd   5.326667  0.569167  1.166667    485.000000
ff   5.377105  1.406316  1.859649    434.649123
h    5.368297  3.416667  3.036232    968.181159
j    3.218750  0.281250  1.500000    132.625000
n    4.593750  1.823750  2.750000    190.750000
o   11.667500  0.167500  0.000000  50060.000000
v    4.352845  1.813960  2.192982    792.992481
z   14.098750  8.218750  6.750000   2254.875000

In [72]:
# Let's see both the standard deviation and the sum of everything in credit_approval_group
grouped.agg(['std', 'sum'])

C                   H                  K                  O  \
          std       sum       std      sum       std  sum           std   
G                                                                         
?    1.924649     9.960  0.178683    0.625  0.000000    0    113.291659   
bb   4.516656   277.245  3.930002  179.050  5.180494  172   6935.782003   
dd   5.131462    31.960  0.910606    3.415  2.041241    7    593.422278   
ff   6.250560   306.495  3.475416   80.160  3.642206  106   1447.721658   
h    5.093527   740.825  3.622352  471.500  4.532783  419   3309.298566   
j    3.526920    25.750  0.795495    2.250  3.854496   12    334.838767   
n    7.005502    18.375  2.388484    7.295  4.272002   11    269.929620   
o   14.612362    23.335  0.236881    0.335  0.000000    0  70625.825305   
v    4.541131  1736.785  2.747279  723.770  5.077327  875   3265.138741   
z    6.087412   112.790  9.189899   65.750  7.592289   54   4578.639099   

            
       sum  
G           
?      341  
bb  103974  
dd    2910  
ff   24775  
h   133609  
j     1061  
n      763  
o   100120  
v   316404  
z    18039

In [75]:
# Let's see the count on column H, the sum on column C, and the mean on column O.
my_funcs = {'H': 'count', 'C': 'sum', 'O': 'mean'}

grouped.agg(my_funcs)

H         C             O
G                              
?     9     9.960     37.888889
bb   59   277.245   1762.271186
dd    6    31.960    485.000000
ff   57   306.495    434.649123
h   138   740.825    968.181159
j     8    25.750    132.625000
n     4    18.375    190.750000
o     2    23.335  50060.000000
v   399  1736.785    792.992481
z     8   112.790   2254.875000

## Merge/join; or, how Pandas can be like SQL

In Pandas, it's possible to combine `DataFrames` and `Series` much like you would in SQL. For the examples in this section, we'll work with smaller `DataFrames` rather than our datasets. It's easier to provide proof of concept this way, as well as explain what's going on

Let's start by appending a row to a `DataFrame`. We can do that by passing in a dictionary to the append function, and setting `ignore_index` equal to `True`.

In [76]:
data = pd.DataFrame({'col1': ['i', 'love', 'pandas', 'so', 'much'],
        'col2': ['so', 'will', 'you', 'i', 'promise']})
data.append({'col1': 'dude', 'col2': 'dude'}, ignore_index=True)

col1     col2
0       i       so
1    love     will
2  pandas      you
3      so        i
4    much  promise
5    dude     dude

Appending a column is also easy. You can do that by setting a new column name equal to a list or a `Series`.

In [77]:
data['col3'] = ['how', 'do', 'you', 'like', 'oscon']
data

col1     col2   col3
0       i       so    how
1    love     will     do
2  pandas      you    you
3      so        i   like
4    much  promise  oscon

However, this will not work if your new column in a different length than the original `DataFrame`.

In [78]:
data['col4'] = ['I', 'am', 'too', 'short']
data

ValueError: Length of values does not match length of index

### Merge

You can `merge()` in different ways, just like joining in SQL. Let's look at an imaginary taco dataset:

In [96]:
tacos = pd.read_csv('../data/tacos.csv')
tacos

name      restaurant  number_of_tacos  score
0   Sarah      Taco Party                4    3.6
1  Georgi      Taco Mania                6    2.5
2   Sammy  Paradise Tacos               10    5.0
3   Peter      Taco Party                8    4.3
4     Rob      Taco Mania                8    3.4

Let's also look at an imaginary taco toppings dataset:

In [97]:
taco_toppings = pd.read_csv('../data/taco_toppings.csv')
taco_toppings

name favorite_topping least_favorite_topping corn_or_flour
0   Sammy            bacon                  slime          corn
1   Peter          avocado                   dirt         flour
2  Georgi         jalapeno                   dirt         flour
3   Sarah           cheese                 celery          corn
4     Rob            salsa                  slime         flour

Notice that we have a unique identifier in each dataset: the name column. We have the same five people. Let's merge these `DataFrames` together. You don't even need to pass the key to merge; `merge()` will automatically infer which key to use based on if it exists in both `DataFrames`. 

In [98]:
tacos.merge(taco_toppings)

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6           cheese   
1  Georgi      Taco Mania                6    2.5         jalapeno   
2   Sammy  Paradise Tacos               10    5.0            bacon   
3   Peter      Taco Party                8    4.3          avocado   
4     Rob      Taco Mania                8    3.4            salsa   

  least_favorite_topping corn_or_flour  
0                 celery          corn  
1                   dirt         flour  
2                  slime          corn  
3                   dirt         flour  
4                  slime         flour

By default, `merge()` performs a left outer join, which means it takes the key from the "left" `DataFrame` - the `DataFrame` that is passed in as the first parameter - and matches the right to it.

Generally speaking, full outer joins will join everything as a union, meaning that everything will be joined even if there are missing values; inner joins will join everything as an intersection, meaning that if a value does not appear in a row in a `DataFrame`, that row will be left out.

Let's look at a couple of other ways of merging. First, let's append a row to our tacos `DataFrame`.

In [100]:
tacos = tacos.append({'name': 'Dan', 'restaurant': 'Tres Carnes', 'number_of_tacos': 7, 'score': 3.8}, ignore_index=True)
tacos

name      restaurant  number_of_tacos  score
0   Sarah      Taco Party                4    3.6
1  Georgi      Taco Mania                6    2.5
2   Sammy  Paradise Tacos               10    5.0
3   Peter      Taco Party                8    4.3
4     Rob      Taco Mania                8    3.4
5     Dan     Tres Carnes                7    3.8

Now, let's do a full outer merge.

In [101]:
tacos.merge(taco_toppings, how='outer')

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6           cheese   
1  Georgi      Taco Mania                6    2.5         jalapeno   
2   Sammy  Paradise Tacos               10    5.0            bacon   
3   Peter      Taco Party                8    4.3          avocado   
4     Rob      Taco Mania                8    3.4            salsa   
5     Dan     Tres Carnes                7    3.8              NaN   

  least_favorite_topping corn_or_flour  
0                 celery          corn  
1                   dirt         flour  
2                  slime          corn  
3                   dirt         flour  
4                  slime         flour  
5                    NaN           NaN

You can see that the entire tacos `DataFrame` has been merged, even though 'Dan' does not exist in the `taco_toppings DataFrame`.

However, if we do the same thing and use a right outer join, we'll only use the keys from the `taco_toppings DataFrame` and Dan will be left out.

In [102]:
tacos.merge(taco_toppings, how='right')

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6           cheese   
1  Georgi      Taco Mania                6    2.5         jalapeno   
2   Sammy  Paradise Tacos               10    5.0            bacon   
3   Peter      Taco Party                8    4.3          avocado   
4     Rob      Taco Mania                8    3.4            salsa   

  least_favorite_topping corn_or_flour  
0                 celery          corn  
1                   dirt         flour  
2                  slime          corn  
3                   dirt         flour  
4                  slime         flour

### Join

The `join()` function gives you a way way to combine `DataFrames` without needing a key. `Taco_extra`, which contains data about chips and spiciness level, has no name column.

In [85]:
taco_extra = pd.read_csv('../data/taco_extra.csv')
taco_extra

chips spiciness
0   yes       hot
1    no      mild
2    no    medium
3   yes       hot
4   yes       hot

It's easy to join this to our taco `DataFrame`.

In [86]:
tacos.join(taco_extra)

name      restaurant  number_of_tacos  score chips spiciness
0   Sarah      Taco Party                4    3.6   yes       hot
1  Georgi      Taco Mania                6    2.5    no      mild
2   Sammy  Paradise Tacos               10    5.0    no    medium
3   Peter      Taco Party                8    4.3   yes       hot
4     Rob      Taco Mania                8    3.4   yes       hot
5     Dan     Tres Carnes                7    3.8   NaN       NaN

You can also specify how to join. The default is outer, but we can change it to inner and Dan will be left out again.

In [87]:
tacos.join(taco_extra, how='inner')

name      restaurant  number_of_tacos  score chips spiciness
0   Sarah      Taco Party                4    3.6   yes       hot
1  Georgi      Taco Mania                6    2.5    no      mild
2   Sammy  Paradise Tacos               10    5.0    no    medium
3   Peter      Taco Party                8    4.3   yes       hot
4     Rob      Taco Mania                8    3.4   yes       hot

It's possible to join more than two `DataFrames` at a time. Let's slice off the name column from taco_toppings.

In [88]:
taco_toppings_noname = taco_toppings.iloc[:, 1:]

taco_toppings_noname

favorite_topping least_favorite_topping corn_or_flour
0            bacon                  slime          corn
1          avocado                   dirt         flour
2         jalapeno                   dirt         flour
3           cheese                 celery          corn
4            salsa                  slime         flour

Joining this frame with tacos and taco_extra is as easy as chaining two joins together. Again, it's all an outer join, so even though there's no toppings or extra data for Dan, he's still included in the `DataFrame`.

In [89]:
tacos.join(taco_toppings_noname).join(taco_extra)

name      restaurant  number_of_tacos  score favorite_topping  \
0   Sarah      Taco Party                4    3.6            bacon   
1  Georgi      Taco Mania                6    2.5          avocado   
2   Sammy  Paradise Tacos               10    5.0         jalapeno   
3   Peter      Taco Party                8    4.3           cheese   
4     Rob      Taco Mania                8    3.4            salsa   
5     Dan     Tres Carnes                7    3.8              NaN   

  least_favorite_topping corn_or_flour chips spiciness  
0                  slime          corn   yes       hot  
1                   dirt         flour    no      mild  
2                   dirt         flour    no    medium  
3                 celery          corn   yes       hot  
4                  slime         flour   yes       hot  
5                    NaN           NaN   NaN       NaN

## Lesson: Let's merge some dataframes!

In [114]:
# Can you merge following DataFrames together?
pizza = pd.read_csv('../data/pizza.csv')
pizza_toppings = pd.read_csv('../data/pizza_toppings.csv')

# Merge them here
pizza.merge(pizza_toppings)

name pizza_style  number_of_slices flavored_crust favorite_topping  \
0     Will    new_york                 5            yes        pepperoni   
1     Lane    new_york                 4             no        anchovies   
2     Dave     chicago                 4             no     green_pepper   
3  Shannon     chicago                 6             no            onion   
4    Ethan     chicago                 3            yes              ham   

  least_favorite  
0      anchovies  
1      pineapple  
2      anchovies  
3      anchovies  
4      bbq_sauce

In [115]:
# Let's join pizza to another dataset, pizza_extra
pizza_extra = pd.read_csv('../data/pizza_extra.csv')

pizza.join(pizza_extra)

name pizza_style  number_of_slices flavored_crust       meal  \
0     Will    new_york                 5            yes     dinner   
1     Lane    new_york                 4             no      lunch   
2     Dave     chicago                 4             no  breakfast   
3  Shannon     chicago                 6             no  breakfast   
4    Ethan     chicago                 3            yes        NaN   

  heating_method  
0           oven  
1      microwave  
2           oven  
3      microwave  
4            NaN

In [116]:
# Let's only join them together where all the data is present
pizza.join(pizza_extra, how='inner')

name pizza_style  number_of_slices flavored_crust       meal  \
0     Will    new_york                 5            yes     dinner   
1     Lane    new_york                 4             no      lunch   
2     Dave     chicago                 4             no  breakfast   
3  Shannon     chicago                 6             no  breakfast   

  heating_method  
0           oven  
1      microwave  
2           oven  
3      microwave

In [117]:
# Can you join all three dataframes together, first by merging pizza and pizza_toppings, then joining that to pizza_extra?
pizza_joined = pizza.merge(pizza_toppings)
pizza_joined.join(pizza_extra)

name pizza_style  number_of_slices flavored_crust favorite_topping  \
0     Will    new_york                 5            yes        pepperoni   
1     Lane    new_york                 4             no        anchovies   
2     Dave     chicago                 4             no     green_pepper   
3  Shannon     chicago                 6             no            onion   
4    Ethan     chicago                 3            yes              ham   

  least_favorite       meal heating_method  
0      anchovies     dinner           oven  
1      pineapple      lunch      microwave  
2      anchovies  breakfast           oven  
3      anchovies  breakfast      microwave  
4      bbq_sauce        NaN            NaN

## Pivoting

You can pivot in Pandas just like you would in Excel. `pivot_table()` takes in four requires parameters: the `DataFrame`, the column to use for the index, the column to use for the columns, and the column to use for the values. `pivot_table()` also has an `aggfunc` parameter that defaults to the mean of the values, but you can pass in other functions, just as we did in the `agg()` function before.

Let's look at the mean weight per model number and number of cylinders combination.

In [118]:
pd.pivot_table(auto_mpg, values='weight', index='model', columns='cylinders')

cylinders     3            4     5            6            8
model                                                       
70          NaN  2292.571429   NaN  2710.500000  3940.055556
71          NaN  2056.384615   NaN  3171.875000  4537.714286
72         2330  2382.642857   NaN          NaN  4228.384615
73         2124  2338.090909   NaN  2917.125000  4279.050000
74          NaN  2151.466667   NaN  3320.000000  4438.400000
75          NaN  2489.250000   NaN  3398.333333  4108.833333
76          NaN  2306.600000   NaN  3349.600000  4064.666667
77         2720  2205.071429   NaN  3383.000000  4177.500000
78          NaN  2296.764706  2830  3314.166667  3563.333333
79          NaN  2357.583333  3530  3025.833333  3862.900000
80         2420  2360.080000  2950  3145.500000          NaN
81          NaN  2275.476190   NaN  3093.571429  3725.000000
82          NaN  2402.321429   NaN  2931.666667          NaN

If a cell contains NaN, it means that that combination doesn't exist within the `DataFrame`.

We can pass in multiple column names to the rows and cols parameters. This creates a `multiindex`.

If we add the origin column to our pivot table, we can look at the average weight of all of the model/origin combinations against the number of cylinders the cars have.

In [119]:
pd.pivot_table(auto_mpg, values='weight', index=['model', 'origin'], columns='cylinders')

cylinders        3            4     5            6            8
model origin                                                   
70    1        NaN          NaN   NaN  2710.500000  3940.055556
      2        NaN  2309.200000   NaN          NaN          NaN
      3        NaN  2251.000000   NaN          NaN          NaN
71    1        NaN  2178.600000   NaN  3171.875000  4537.714286
      2        NaN  2024.000000   NaN          NaN          NaN
      3        NaN  1936.000000   NaN          NaN          NaN
72    1        NaN  2263.800000   NaN          NaN  4228.384615
      2        NaN  2573.200000   NaN          NaN          NaN
      3       2330  2293.000000   NaN          NaN          NaN
73    1        NaN  2355.500000   NaN  2932.857143  4279.050000
      2        NaN  2335.714286   NaN          NaN          NaN
      3       2124  2329.000000   NaN  2807.000000          NaN
74    1        NaN  2372.666667   NaN  3320.000000  4438.400000
      2        NaN  2139.333333   NaN          NaN          NaN
      3        NaN  2053.000000   NaN          NaN          NaN
75    1        NaN  2615.500000   NaN  3398.333333  4108.833333
      2        NaN  2571.166667   NaN          NaN          NaN
      3        NaN  2303.250000   NaN          NaN          NaN
76    1        NaN  2318.200000   NaN  3343.250000  4064.666667
      2        NaN  2438.285714   NaN  3820.000000          NaN
      3        NaN  1980.000000   NaN  2930.000000          NaN
77    1        NaN  2346.000000   NaN  3525.000000  4177.500000
      2        NaN  2138.750000   NaN          NaN          NaN
      3       2720  2060.000000   NaN  2815.000000          NaN
78    1        NaN  2417.500000   NaN  3322.000000  3563.333333
      2        NaN  2256.666667  2830  3275.000000          NaN
      3        NaN  2221.250000   NaN          NaN          NaN
79    1        NaN  2435.857143   NaN  3025.833333  3862.900000
      2        NaN  2415.000000  3530          NaN          NaN
      3        NaN  1997.500000   NaN          NaN          NaN
80    1        NaN  2729.333333   NaN  3381.000000          NaN
      2        NaN  2272.750000  2950          NaN          NaN
      3       2420  2222.181818   NaN  2910.000000          NaN
81    1        NaN  2330.625000   NaN  3166.250000  3725.000000
      2        NaN  2580.000000   NaN  3160.000000          NaN
      3        NaN  2140.000000   NaN  2915.000000          NaN
82    1        NaN  2585.882353   NaN  2931.666667          NaN
      2        NaN  2055.000000   NaN          NaN          NaN
      3        NaN  2132.777778   NaN          NaN          NaN

You can apply different aggregate functions to a pivot table. Let's look at the total weight per model/cylinder combination.

In [120]:
pd.pivot_table(auto_mpg, values='weight', index='model', columns='cylinders', aggfunc='sum')

cylinders     3      4     5      6      8
model                                     
70          NaN  16048   NaN  10842  70921
71          NaN  26733   NaN  25375  31764
72         2330  33357   NaN    NaN  54969
73         2124  25719   NaN  23337  85581
74          NaN  32272   NaN  23240  22192
75          NaN  29871   NaN  40780  24653
76          NaN  34599   NaN  33496  36582
77         2720  30871   NaN  16915  33420
78          NaN  39045  2830  39770  21380
79          NaN  28291  3530  18155  38629
80         2420  59002  2950   6291    NaN
81          NaN  47785   NaN  21655   3725
82          NaN  67265   NaN   8795    NaN

## Lesson: let's pivot!

In [125]:
# Create a pivot_table for credit_approval with column A as the index, column J as the columns, and column H as the values.
pd.pivot_table(df, values='H', index='A', columns='J')

J         f         t
A                    
?  0.750000  2.416667
a  1.405370  2.188725
b  1.673291  3.578658

In [126]:
# Now, change the aggfunc to the standard deviation.
pd.pivot_table(df, values='H', index='A', columns='J', aggfunc='std')

J         f         t
A                    
?  1.203482  2.036132
a  2.316387  2.736864
b  2.660253  4.531032

In [ ]:
# Finally, can you come up with your own pivot_table?


# For those using IPython Notebook/Wakari/NBViewer: Go to the [data_analysis](data_analysis.ipynb) notebook!

# For those using code files, go to data_analysis.py!